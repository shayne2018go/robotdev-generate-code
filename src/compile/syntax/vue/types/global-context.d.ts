declare namespace GlobalContext {
  // export interface Router {
  //   id: string; // 页面id
  //   name?: string; // name
  //   path: string; // path
  //   filePath: string; // 文件地址
  // }

  /**
   * 页面
   */
  export interface Page {
    id: string;
    key: string;
    routerName?: string; // vur router 中的name
    routerPath: string; //  vur router 中的path
    source: {
      filePath: string; // 文件地址
    };
    protocol: CodeSchema.Page; // 协议
  }

  /**
   * 组件
   */
  export interface Component {
    id: string; // tagId
    key: string; // tag 标签
    source?: {
      filePath?: string; // 本地路径
      packageName?: string; // 包路径
      exportName: string; // 导出名 (不能为空 默认default ) aaa
      alias?: string; // 别名
    };
    protocol: CodeSchema.Component; // 协议
  }

  /**
   * 函数
   */
  export interface Function {
    id: string; // 函数id
    key: string; // 函数名称
    source?: {
      filePath?: string; // 本地路径
      packageName?: string; // 包路径
      exportName: string; // 导出名 (不能为空 默认default )
      alias?: string; // 别名
    };
    protocol: CodeSchema.Function_Protocol; // 协议
  }

  /**
   * 行为
   */
  export interface Action {
    id: string; // 行为id
    key: string; // 行为名称
    source?: {
      filePath?: string; // 本地路径
      packageName?: string; // 包路径
      exportName: string; // 导出名 (不能为空 默认default )
      alias?: string; // 别名
    };
    protocol: CodeSchema.Action_Protocol; // 协议
  }

  export interface Api {
    id: string; // 行为id
    key: string; // 行为名称
    source: {
      filePath?: string; // 本地路径
      packageName?: string; // 包路径
      exportName: string; // 导出名 (不能为空 默认default )
      alias?: string; // 别名
    };
    protocol: CodeSchema.Api_Protocol; // 协议
  }

  export interface Event {
    id: string; // 行为id
    key: string; // 行为名称
    source?: {
      filePath?: string; // 本地路径
      packageName?: string; // 包路径
      exportName: string; // 导出名 (不能为空 默认default )
      alias?: string; // 别名
    };
    protocol: CodeSchema.Event_Protocol; // 协议
  }

  export interface Property {
    id: string; // 属性id
    key: string; // 属性名称
    protocol: CodeSchema.Property_Protocol; // 协议
  }
}
