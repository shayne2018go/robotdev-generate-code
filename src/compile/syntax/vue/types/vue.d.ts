import { ICS_ACTION } from '@/types/action';
import { ICS_Api } from '@/types/api';
import { ICS_Component } from '@/types/component';
import { ICS_Function } from '@/types/function';

declare namespace VueTypes {
  /**
   * 组件
   */
  export interface Component {
    id: string; // tagId
    key: string; // tag 标签
    source?: {
      filePath?: string; // 本地路径
      packageName?: string; // 包路径
      exportName: string; // 导出名 (不能为空 默认default ) aaa
      alias?: string; // 别名
    };
    protocol: ICS_Component; // 协议
  }

  /**
   * 函数
   */
  export interface Function {
    id: string; // 函数id
    key: string; // 函数名称
    source: {
      filePath?: string; // 本地路径
      packageName?: string; // 包路径
      exportName: string; // 导出名 (不能为空 默认default )
      alias?: string; // 别名
    };
    protocol: ICS_Function; // 协议
  }

  /**
   * 行为
   */
  export interface Action {
    id: string; // 行为id
    key: string; // 行为名称
    source: {
      filePath?: string; // 本地路径
      packageName?: string; // 包路径
      exportName: string; // 导出名 (不能为空 默认default )
      alias?: string; // 别名
    };
    protocol: ICS_ACTION; // 协议
  }

  export interface Api {
    id: string; // 行为id
    key: string; // 行为名称
    source: {
      filePath?: string; // 本地路径
      packageName?: string; // 包路径
      exportName: string; // 导出名 (不能为空 默认default )
      alias?: string; // 别名
    };
    protocol: ICS_Api; // 协议
  }
}
