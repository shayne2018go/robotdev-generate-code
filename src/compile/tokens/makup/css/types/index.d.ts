import { CssDataType } from './cssDataType';

export type IBlock_DataType = keyof typeof CssDataType;

export type IBlock_DataTypeEnum = typeof CssDataType;

/**
 * 块状代码schema
 */
export interface IBlock_BlockSchema {
  key: string;
  value: Array<IBlock_Item>;
}

/**
 * 样式schema
 */
export interface IBlock_AttrSchema {
  key: string;
  value: string;
}

/**
 * @ 声明的 schema
 */
export interface IBlock_CommonSchema {
  key: string;
  value: string;
}

export type IBlock = Array<IBlock_Item>;

export type IBlock_Item = IBlock_BlockItem | IBlock_AttrItem | IBlock_CommonItem;

export interface IBlock_BlockItem extends IBlock_BlockSchema {
  type: IBlock_DataTypeEnum['block'];
}

export interface IBlock_AttrItem extends IBlock_AttrSchema {
  type: IBlock_DataTypeEnum['attr'];
}

export interface IBlock_CommonItem extends IBlock_CommonSchema {
  type: IBlock_DataTypeEnum['common'];
}

export type IAttr = Array<IBlock_AttrSchema>;
