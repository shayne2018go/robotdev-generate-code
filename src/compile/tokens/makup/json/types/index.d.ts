import { JsonDataType } from './jsonDataType';

export type IJson_DataType = keyof typeof JsonDataType;

export type IJson_DataTypeEnum = typeof JsonDataType;

export interface IJson_Object {
  type: IJson_DataTypeEnum['object'];
  value: Array<IJson_Object_Item>;
}

export interface IJson_Object_Item {
  key: string;
  value: IJson;
}

export interface IJson_Array {
  type: IJson_DataTypeEnum['array'];
  value: Array<IJson>;
}

export interface IJson_String {
  type: IJson_DataTypeEnum['string'];
  value: string;
}

export interface IJson_Number {
  type: IJson_DataTypeEnum['number'];
  value: number;
}

export interface IJson_Boolean {
  type: IJson_DataTypeEnum['boolean'];
  value: boolean;
}

export interface IJson_Null {
  type: IJson_DataTypeEnum['null'];
}

export type IJson = IJson_Null | IJson_String | IJson_Number | IJson_Boolean | IJson_Object | IJson_Array;

/**
 * null string number boolean 为简单类型数据
 */
export type ISimpleData = IJson_Null | IJson_String | IJson_Number | IJson_Boolean;


/**
 * object array 为复杂类型数据
 */
export type IComplexData = IJson_Object | IJson_Array;