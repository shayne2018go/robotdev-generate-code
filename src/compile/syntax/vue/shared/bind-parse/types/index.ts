import {
  ArrayExpression,
  AssignmentExpression,
  BinaryExpression,
  BooleanLiteral,
  CallExpression,
  DecimalLiteral,
  Identifier,
  IfStatement,
  MemberExpression,
  NullLiteral,
  NumericLiteral,
  ObjectExpression,
  OptionalMemberExpression,
  StringLiteral,
} from '@babel/types';
import { CompileCurrentCtx } from '../../../compilePages';
import { genVarName } from '../../helper';

export type ActionAst = CallExpression | AssignmentExpression;
export type ActionsAst = (CallExpression | AssignmentExpression)[];

export type BindAst =
  | OptionalMemberExpression
  | MemberExpression
  | ArrayExpression
  | StringLiteral
  | NumericLiteral
  | BooleanLiteral
  | ObjectExpression
  | DecimalLiteral
  | CallExpression
  | NullLiteral
  | Identifier
  | BinaryExpression;

export type BindRdData =
  | CodeSchema.DataValue_GetVar
  | CodeSchema.DataValue_GetApiData
  | CodeSchema.DataValue_GetParam
  | CodeSchema.DataValue_GetEventData
  | CodeSchema.DataValue_GetSlotData
  | CodeSchema.DataValue_GetEachData
  | CodeSchema.DataValue_GetArguments
  // | CodeSchema.DataValue_GetModelData
  | CodeSchema.DataValue_GetCmptPropData
  | CodeSchema.DataValue_TableData;

export type LiteralAst =
  | ArrayExpression
  | StringLiteral
  | NumericLiteral
  | BooleanLiteral
  | ObjectExpression
  | DecimalLiteral
  | CallExpression
  | Identifier;

export type BindParseCtx<T extends CodeSchema.Page | CodeSchema.Component = CodeSchema.Page> = {
  global: CompileCurrentCtx['global'];
  scope: CompileCurrentCtx['scope'] & {
    node?: CodeSchema.ComponentNode;
    prop?: CodeSchema.Property;
    event?: CodeSchema.Event;
    actions?: {
      genVarName: ReturnType<typeof genVarName>;
      map: {
        [id: CodeSchema.Action['id']]: {
          data: CodeSchema.Action;
          protocol: CodeSchema.Action_Protocol;
          varName: string;
          parametersVarNames: {
            [id: CodeSchema.Action_Protocol['parameters'][number]['id']]: {
              protocol: CodeSchema.Action_Protocol['parameters'][number];
              varName: string;
              memberVarNames: {
                [id: CodeSchema.Property_Protocol['id']]: {
                  protocol: CodeSchema.Property_Protocol;
                  varName: string;
                };
              };
            };
          };
        };
      };
    };
  };
};

export type AccessPath = {
  type: 'object' | 'array';
  key: string | number;
}[];

export type SplitProps = {
  key: string;
  value: BindAst | undefined;
}[];

export type ReturnRef = {
  type: 'split' | 'ast' | 'error';
  value?: ActionAst | ActionsAst | IfStatement | BindAst | SplitProps | undefined;
};
