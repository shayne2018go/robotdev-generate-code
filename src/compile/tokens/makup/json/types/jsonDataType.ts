export const JsonDataType = {
  null: 'null',
  string: 'string',
  number: 'number',
  boolean: 'boolean',
  array: 'array',
  object: 'object',
} as const;