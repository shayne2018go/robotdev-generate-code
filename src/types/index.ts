export * from './ICodeSchema';

export * from './DBSchema';

export * from './DBWSchema';
