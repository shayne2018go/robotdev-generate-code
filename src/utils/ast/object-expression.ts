import {
  ObjectExpression,
  Expression,
  ObjectProperty,
  ObjectMethod,
  SpreadElement,
  objectExpression,
  objectProperty,
  identifier,
  booleanLiteral,
  stringLiteral,
  numericLiteral,
  arrayExpression,
  functionExpression,
  callExpression,
} from '@babel/types';
import { isArray, isPlanObject } from '../is';
import { watch } from 'fs-extra';
import { d } from 'vitest/dist/types-e3c9754d';
import { generate } from '@/generate';

function getObjectExpressionKeys(node: ObjectExpression): string[] {
  const keys: string[] = [];
  for (const prop of node.properties) {
    if ((prop.type === 'ObjectProperty' || prop.type === 'ObjectMethod') && !prop.computed) {
      if (prop.key.type === 'Identifier') {
        keys.push(prop.key.name);
      } else if (prop.key.type === 'StringLiteral') {
        keys.push(prop.key.value);
      }
    }
  }
  return keys;
}

function getObjectPropertyByKey(node: ObjectExpression, key: string | string[]): ObjectProperty | null {
  const path = Array.isArray(key) ? key : key.split('.');
  let current: ObjectProperty | ObjectExpression | null = node;

  while (path.length) {
    const curKey = path.pop();
    if (!curKey) {
      current = null;
      break;
    }
    const objectKeys = getObjectExpressionKeys(node);
    if (objectKeys.includes(curKey)) {
      if (!current) {
        break;
      }
      if (current.type === 'ObjectExpression') {
        for (const prop of current.properties) {
          if (prop.type === 'ObjectProperty' && !prop.computed) {
            current = prop;
          }
        }
      }
    } else {
      current = null;
      break;
    }
  }
  return current as ObjectProperty | null;
}

function updateObjectExpression(node: ObjectExpression, key: string | string[], value: any) {
  const path = Array.isArray(key) ? key : key.split('.');
  let current: any = node;

  let pathIndex = 1;
  let updatePath = path.slice(0, pathIndex);

  while (updatePath.length < path.length) {
    debugger;

    const isExit = getObjectPropertyByKey(node, updatePath);

    pathIndex++;
    updatePath = path.slice(0, pathIndex);
    if (isExit) {
      continue;
    }
  }

  // while (path.length) {
  //   const curKey = path.shift();
  //   if (!curKey) {
  //     break;
  //   }
  //   const objectKeys = getObjectExpressionKeys(node);
  //   if (objectKeys.includes(curKey)) {
  //     if (current.type === 'ObjectExpression') {
  //       for (const prop of current.properties) {
  //         if (prop.type === 'ObjectProperty' && !prop.computed) {
  //           current = prop;
  //         }
  //       }
  //     }
  //   } else {
  //     current = createExpression(path, value);
  //     break;
  //   }
  // }
}

function createExpression(path: string[], value: any) {
  let current = null;

  const key = path.shift();

  if (!key) {
    return null;
  }

  if (path.length) {
    const property = createExpression(path, value);
    if (property) {
      if (property.type === 'ObjectExpression') {
        current = objectExpression([objectProperty(identifier(key), property)]);
      } else if (property.type === 'ObjectProperty') {
        current = objectExpression([property]);
      }
    }
  } else {
    current = objectProperty(identifier(key), getValueAstExpressionByType(value));
  }
  return current;
}

function getValueAstExpressionByType(value: unknown): any {
  if (typeof value === 'string') {
    return stringLiteral(value);
  } else if (typeof value === 'number') {
    return numericLiteral(value);
  } else if (typeof value === 'boolean') {
    return booleanLiteral(value);
  } else if (isArray(value)) {
    return arrayExpression(value.map(getValueAstExpressionByType));
  } else if (isPlanObject(value)) {
    const objectProperties = [] as (ObjectMethod | ObjectProperty | SpreadElement)[];
    Object.keys(value).forEach((key) => {
      const keyValue = value[key];
      const propertyValue = getValueAstExpressionByType(keyValue);
      objectProperties.push(objectProperty(identifier(key), propertyValue));
    });
    return objectExpression(objectProperties);
  }
  return null;
}

export { createExpression, updateObjectExpression };
