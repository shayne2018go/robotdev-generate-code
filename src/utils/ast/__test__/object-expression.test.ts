import { describe, expect, it } from 'vitest';
import { createExpression, updateObjectExpression } from '../object-expression';
import { ObjectExpression } from '@babel/types';

describe('object-expression', () => {
  it('createExpression', () => {
    const expression = createExpression(['a', 'b'], { d: 123 });

    expect(expression).toMatchSnapshot();
  });

  it('updateObjectExpression', () => {
    const expression = createExpression(['a', 'b'], { d: 123 });

    if (expression) {
      updateObjectExpression(expression as ObjectExpression, 'b.e', true);
    }
  });
});
