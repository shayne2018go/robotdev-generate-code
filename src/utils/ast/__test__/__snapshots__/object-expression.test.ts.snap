// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`object-expression > createExpression 1`] = `
{
  "properties": [
    {
      "computed": false,
      "decorators": null,
      "key": {
        "name": "b",
        "type": "Identifier",
      },
      "shorthand": false,
      "type": "ObjectProperty",
      "value": {
        "properties": [
          {
            "computed": false,
            "decorators": null,
            "key": {
              "name": "d",
              "type": "Identifier",
            },
            "shorthand": false,
            "type": "ObjectProperty",
            "value": {
              "type": "NumericLiteral",
              "value": 123,
            },
          },
        ],
        "type": "ObjectExpression",
      },
    },
  ],
  "type": "ObjectExpression",
}
`;
